cowbird/handlers/impl/geoserver.py

```diff
@@ -126,11 +134,120 @@ def user_deleted(self, user_name):
+            res_type = res_info["resource_type"]
+            if res_type == "layer":
+                layer_name = res_info["resource_name"]
+
+            self._update_resource_files_permissions(resource_type=res_type,
+                                                    permission=permission,
+                                                    resource_id=res_id,
+                                                    workspace_name=workspace_name,
+                                                    layer_name=layer_name)
+            if res_type == "workspace":
+                self._update_res_children_files_permissions(children_res_tree=res_info["children"],
+                                                            permission=permission,
+                                                            workspace_name=workspace_name,
+                                                            layer_name=layer_name)
+
+    def _update_permissions_on_filesystem(self, permission):
+        # type: (Permission) -> None
+        """
+        Updates the permissions of files on the file system, after receiving a permission webhook event from Magpie.
+        """
+        if permission.name not in LAYER_READ_PERMISSIONS + LAYER_WRITE_PERMISSIONS:
+            LOGGER.info("Nothing to do, since it is not a permission for a Geoserver resource.")
+            return
+
+        magpie_handler = HandlerFactory().get_handler("Magpie")
+
+        if permission.user is None:
+            raise NotImplementedError("A permission change on a group is not supported for now on Geoserver, since"
+                                      "workspaces are based on users only.")
+        workspace_name = permission.user
+
+        # Get resource type with resource id
+        resource_tree = magpie_handler.get_parents_resource_tree(permission.resource_id)
+        resource_type = resource_tree[-1]["resource_type"]
+
+        if resource_type in ["workspace", "layer"]:
+            layer_name = permission.resource_full_name.split("/")[-1] if resource_type == "layer" else None
+            self._update_resource_files_permissions(resource_type=resource_type,
+                                                    permission=permission,
+                                                    resource_id=permission.resource_id,
+                                                    workspace_name=workspace_name,
+                                                    layer_name=layer_name)
+
+        if resource_type in ["service", "workspace"] and permission.scope == "recursive":
+            children_res_tree = magpie_handler.get_children_resource_tree(permission.resource_id)
+            self._update_res_children_files_permissions(children_res_tree=children_res_tree,
+                                                        permission=permission,
+                                                        workspace_name=workspace_name)
```

Comment:
I think I would leave getting the children resource tree outside of the function here, to avoid sending multiple requests to Magpie to get the children tree, unless I misunderstood what you meant here? 
I request the tree one time, then apply the recursive function to it.

Reply:
Hard to tell with simply the static analysis, but my intuition tells me that if the `self._update_resource_files_permissions` call (with the same if condition) is moved inside `self._update_res_children_files_permissions`, then you could simply call `self._update_res_children_files_permissions` with the parent resource (the one passed in `self._update_resource_files_permissions`, and the recursive handling will automatically perform the nested calls as necessary. In this part of the code, you wouldn't need to worry about handling the resource/children separately, you just pass the resource directly to the recursive method and it does the job.

If there is a more intricate condition/situation that I'm not seeing and would make this refactor more complicated, then the current implementation is still fine since it is called only here. If more equivalent uses spring up later though, the logic will have to be duplicated.